ingest/clusters.py

```diff
@@ -0,0 +1,80 @@
+import copy
+from typing import Dict, Generator, List, Tuple, Union
+
+from ingest_files import IngestFiles
+
+
+class Clusters(IngestFiles):
+
+    ALLOWED_FILE_TYPES = ['text/csv',
+                          'text/plain', 'text/tab-separated-values']
+    MAX_THRESHOLD = 100_000
+    SUBSAMPLE_THRESHOLDS = [MAX_THRESHOLD, 20000, 10000, 1000]
+    COLLECTION_NAME = 'clusters'
+    SUBCOLLECTION_NAME = 'data'
+
+    def __init__(self, file_path, *, name: str = None, study_accession: str = "", domain_ranges=None):
+
+        IngestFiles.__init__(self, file_path, self.ALLOWED_FILE_TYPES)
+        self.header = self.get_next_line(increase_line_count=False)
+        # Second line in cluster is metadata_type
+        self.metadata_types = self.get_next_line(increase_line_count=False)
+        self.unique_values = []
+        self.source_file_type = 'cluster'
+        # self.points = amount of rows
+        self.top_level_doc = {
+            'name': name,
+            'study_accession': study_accession,
+            'source_file_type': 'cluster',
+            'domain_ranges': domain_ranges,
+            'points': self.amount_of_lines,
+        }
+        self.annotation_subdocs = self.create_annotation_subdocs()
+        # sample size needs to be smaller than amount of points
+        # for each cell annotations if there are any
+
+    def update_points(self):
+        self.top_level_doc['points'] = self.amount_of_lines
+
+    def transform(self, rows):
+        """ Add data from cluster files into annotation subdocs in cluster data model"""
+        for idx, column in enumerate(rows):
+            if idx != 0:
+                if self.metadata_types[idx].lower() == 'numeric':
+                    column = round(float(column), 3)
+                elif self.metadata_types[idx].lower() == 'group':
+                    if column not in self.unique_values:
+                        self.unique_values.append(column)
+            annotation = self.header[idx].lower()
```

Comment:
I understand why we `self.metadata_types[idx].lower()`, but this use of `lower()` seems like it would unexpectedly change user data -- e.g. cast "GABAergic cells" to "gabaergic cells".  So I'd suggest:

```suggestion
            annotation = self.header[idx]
```

Reply:
This is the types header row, not the top one (e.g. 2nd line).  We definitely want to cast these to lower case, as was discovered in [SCP-1729](https://broadworkbench.atlassian.net/browse/SCP-1729).  If we don't, and someone supplies a type header value of 'Group' or 'Numeric', it causes annotation values to be handled incorrectly.

Reply:
I retract this - I was looking at ``self.metadata_types[idx]``, and not ``annotation = self.header[idx].lower()``.  Eric is correct - we do not want to downcast the value of the field header.  We want to preserve that case.

Reply:
Resolved in 60db95d